frontend/src/metabase/query_builder/components/DimensionList.jsx

```diff
@@ -31,41 +31,51 @@ export default class DimensionList extends Component {
   };
 
   UNSAFE_componentWillMount() {
-    this._updateSections(this.props.sections);
+    this.updateSections(this.props.sections);
   }
+
   UNSAFE_componentWillReceiveProps(nextProps) {
     if (this.props.sections !== nextProps.sections) {
-      this._updateSections(nextProps.sections);
+      this.updateSections(nextProps.sections);
     }
   }
-  _updateSections(sections) {
-    this.setState({
-      sections: (sections || []).map(section => ({
-        ...section,
-        items: section.items.map(item => ({
-          ...item,
-          name: item.name || (item.dimension && item.dimension.displayName()),
-          icon: item.icon || (item.dimension && item.dimension.icon()),
-        })),
-      })),
-    });
+
+  buildUpdatedSectionItems(items) {
+    return items.map(item => ({
+      ...item,
+      name: item.name || item.dimension?.displayName(),
+      icon: item.icon || item.dimension?.icon(),
+    }));
   }
 
-  _getDimensions() {
+  updateSections(sections = []) {
```

Comment:
I personally think this is harder to understand since there're a lot more variables we need to keep track of to understand the whole picture of `updateSection`.

Previously, we know that the only argument of the function is `sections` and that's it. But now we have to know
- sectionsForNewState
- buildUpdateSectionItems()
- newState

I don't think the previous code, which inlines everything, is that bad. What do you think?

Reply:
Yeah, I can see your point of view here.

Iterated by reverting, but keeping:

1. `(sections = [])` in the argument of the function
2. optional chaining

please let me know if you see improvements to be made here.

Reply:
@gusaiani sounds good on both 👍